torchtext/data/utils.py

```diff
@@ -224,6 +233,74 @@ def _get_ngrams(n):
             yield ' '.join(x)
 
 
+def generate_sp_tokenizer(filename, vocab_size=20000,
+                          model_type="unigram",
+                          model_prefix='m_user'):
+    """Train a SentencePiece tokenizer.
+
+    Arguments:
+        filename: the data file for training SentencePiece model.
+        vocab_size: the size of vocabulary (Default: 20,000).
+        model_type: the type of SentencePiece model, including unigram,
+            bpe, char, word.
+        model_prefix: the prefix of the files saving model and vocab.
+
+    Outputs:
+        The model and vocab are saved in two separate files with
+            model_prefix.
+
+    Examples:
+        >>> from torchtext.data.utils import generate_sp_tokenizer
+        >>> generate_sp_tokenizer('test.csv', vocab_size=23456,
+        >>>                       model_prefix='spm_user')
+        >>> import sentencepiece as spm
+        >>> sp_user = spm.SentencePieceProcessor()
+        >>> sp_user.load('spm_user.model')
+    """
+
+    import sentencepiece as spm
+    spm_training_string = "--input={} \
+                           --vocab_size={} \
+                           --model_prefix={} \
+                           --model_type={}".format(filename,
+                                                   vocab_size,
+                                                   model_prefix,
+                                                   model_type)
+    spm.SentencePieceTrainer.train(spm_training_string)
+    return None
+
+
+def spm_data_generator(spm_name, iterator):
+    """Tokenize and numericalize text sentences with SentencePiece
+       tokenizer and a data iterator.
+
+    Arguments:
+        spm_name: the filename of the sentencepiece model.
+        iterator: an iterator returns the text sentences.
+
+    Outputs:
+        data: a list of integer lists saving the numericalized tokens.
+
+    Examples:
+        >>> from torchtext.data.utils import spm_data_generator
+        >>> iterator = ['Generic data loaders, abstractions, and iterators',
+        >>>             'Pre-built loaders for common NLP datasets']
+        >>> spm_data_generator('spm_user.model', iterator)
+        >>> [[3046, 7706, 319, 14246, 174, 5, 4838, 4, 18995, 4, 5, 14, 41, 86, 6106],
+        >>>  [3493, 15, 14528, 14246, 174, 18, 3054, 3397, 264, 319, 17493]]
+    """
+
+    import sentencepiece as spm
+    sp_user = spm.SentencePieceProcessor()
+    sp_user.load(spm_name)
+
+    data = []
+    for corpus in iterator:
+        token_ids = sp_user.encode_as_ids(corpus)
+        data.append(token_ids)
```

Comment:
This consumes an iterator, yet returns a list which materializes all data. Maybe this should be an iterator